refactor(components,app,protocol-designer): Pull labware transforms out of <Module> and prepare them for labware schema 3

SyntaxColoring opened the pull request:
## Overview

Another piece towards EXEC-1278. This refactors the way the frontend renders things aligned to modules (usually labware aligned to modules). It also updates many of those places so they will work properly when given a labware in schema 3.

## Changelog

Before, we had a pattern like this:

```jsx
<!-- Renders a module SVG. -->
<Module ...> 
  <!-- Children are rendered at the (-x, -y) corner of the "slot" on top of the module. -->
  
  <!-- Renders a labware SVG in the correct spot atop the module. -->
  <LabwareRender ... />

  <!-- e.g. this could be an "edit labware" button. -->
  <SomeUIControl ... />
</Module>
```

With this PR, that changes to this:

```jsx
<Module childrenPositioningMode="passThrough"...>
  <AlignLabwareToModule ...>
    <LabwareRender ... />
  </AlignLabwareToModule>

  <AlignControlToModule ...>
    <SomeUIControl ... />
  </AlignControlToModule />
</Module>
```

Basically, we're taking an SVG transform that used to be built into `<Module>` and moving it into its own component. The new helper components are separated according to their use case:

* `<AlignLabwareToModule>` goes from the module origin to the labware origin, and it works with labware schema 3. When we implement the TypeScript code to, for example, align well A1 of a well plate to well A1 of a Thermocycler, it will automatically benefit from that.
* `<AlignControlToModule>` reimplements the prior behavior: it goes from the module origin to the (-x, -y) corner of the "slot" on top of the module. This is intended to support various non-labware things, like buttons and info overlays, that are rendered in the module's "labware area."

The old behavior is preserved, for now, if you do `<Module childrenPositioningMode="offsetToSlot">`. This is basically equivalent to `<AlignControlToModule>` being built into `<Module>`.

This PR also updates `BaseDeck` to use `childrenPositioningMode="passThrough"`, which means its API changes too. Its callers need to use one of the helper components now.

## Rationale

A recurring theme in our recent work is: when we want to align things to each other, we need the whole picture. To align a labware to a module, we need to know about the module _and_ the labware, both at once. A `<LabwareRender>` can't position itself relative to a module knowing only the labware's definition, and a `<Module>` can't position a labware relative to it using only the module's definition. In principle, this was always true, but we often got away with it. Labware schema 3 doesn't let us get away with it anymore.

Also, a `<Module>` may have multiple child components, and each of those might need its own positioning logic. A schema 2 `<LabwareRender>`, a schema 3 `<LabwareRender>`, and a button will all want to work differently. So we need to be able to mix and match the positioning logic granularly.

The `<AlignLabwareToModule>`+`<AlignControlToModule>` combo is basically the only thing I could come up with that solves both of those.

Alternatives considered:

* Provide more contextual information to the child component, like a `moduleDef` prop: `<LabwareRender labwareDef=... moduleDef=... />`.
  * This isn't great separation of responsibilities.
  * There would also need to be more props than you might think: not just the `moduleDef`, but also the `deckDef` and `slotId`.
  * There are also more child components to modify than you might think: not just `<LabwareRender>`, but also various buttons and overlays.
* Provide more contextual information to the parent `<Module>` component, like `<Module moduleDef=... labwareDef=...`>.
  * This isn't sufficient because it doesn't let us mix and match the positioning logic granularly between different children.
* Instead of building up a nested hierarchy of SVG transforms, calculate the absolute position of everything up-front, in TypeScript, and render everything in one flat pool.
  * I think this is actually a good idea and we should try to do it, but it was too big of a structural difference for me to attempt now.

## Test Plan and Hands on Testing

* [x] Audit all call sites of `<Module>`. If they contain one of the new helper components, they should use `positioningMode="passThrough"`. Otherwise, they should use `positioningMode="offsetToSlot"`.
* [x] Audit all uses of `<BaseDeck>`'s `moduleChildren` prop, which is effectively an indirect call site of `<Module>`. Make sure they all contain one of the new helper components.
* [x] Explore various module visualizations in the various frontends and make sure there are no flagrant problems.

## Review requests

### Big stuff

* This is a churny API change for `<Module>`. Fundamentally, are we okay with the new API?
* Any better name ideas than `AlignControlToModule`? 

### Code-level stuff

* A few of these call sites needed more information than they already had, like maybe they needed the deck ID. I did my best to pull that information from the nearest place that had it, but I often had trouble figuring that out. Glance through the code to see if there's a cleaner way I could be doing any of that.

## Risk assessment

High. Big surface area, lots of little places that we need to remember to update.

The impact of forgetting to update one of those places is that the labware will be shown in the wrong place.

SyntaxColoring committed:
Refactor labware-to-module alignment out of the <Module> component.

* Extract labware-to-module alignment to separate components external to <Module>.
* Use existing utils in shared-data instead of implementing the math ad-hoc inside the component. This also gives us proper labware schema 3 support.
* Refactor the remaining math in <Module> to avoid `let`.
* Drop offset "clamping."

SyntaxColoring committed:
Readability refactor: Some minor deduplication.

SyntaxColoring committed:
Document <Module> x and y props.

SyntaxColoring committed:
Improve getRunModuleRenderInfo() typing.

These props are nullable when they're passed to <Module> because <Module> needs to support the case where it's being rendered in isolation, not atop a deck. This util, on the other hand, needs to guarantee that it outputs non-null, because it *is* used for aligning modules to decks, and we need these props to compute the transforms to do that.

SyntaxColoring committed:
Minor readability refactor.

SyntaxColoring committed:
Expand documentation for labwareChildren and moduleChildren.

This reflects the new semantics (after this PR).

SyntaxColoring committed:
Update <Module>'s call sites.

SyntaxColoring committed:
Update <BaseDeck> moduleChildren call sites.

SyntaxColoring committed:
Merge branch 'edge' into frontend_schema_3_labware_in_modules

Resolve conflicts in:
app/src/organisms/ErrorRecoveryFlows/shared/TwoColLwInfoAndDeck.tsx
app/src/organisms/InterventionModal/MoveLabwareInterventionContent.tsx
components/src/hardware-sim/BaseDeck/BaseDeck.tsx

SyntaxColoring committed:
Merge branch 'edge' into frontend_schema_3_labware_in_modules

Resolve conflicts in app/src/organisms/Desktop/Devices/ProtocolRun/SetupLabware/SetupLabwareMap.tsx.

mjhuff reviewed (approved):
Nice, looks good and makes sense to me. I think we are most likely safe to hold off on merging until the 8.6 branch is cut, but it may not be a bad idea to clear that with @smb2268 / @TamarZanzouri first.

SyntaxColoring committed:
Merge branch 'edge' into frontend_schema_3_labware_in_modules

Resolve conflicts in components/src/hardware-sim/Module/index.tsx

SyntaxColoring committed:
Implement childrenPositioningMode.

SyntaxColoring committed:
Merge fixup: Set childrenPositioningMode="passThrough" where it makes sense

SyntaxColoring committed:
Merge fixup: Rely on childrenPositioningMode="offsetToSlot" where it makes sense

SyntaxColoring committed:
Lint.

SyntaxColoring committed:
Remove unnecessary diff.

SyntaxColoring commented on components/src/hardware-sim/Module/index.tsx:
```diff
@@ -178,29 +176,17 @@ export const Module = (props: Props): JSX.Element => {
       ? 'rotate(0, 0, 0)'
       : `rotate(180, ${rotationCenterX}, ${rotationCenterY})`
 
-  // labwareOffset values are more accurate than our SVG renderings, so ignore any deviations under a certain threshold
-  const clampedLabwareOffsetX =
-    Math.abs(nestedLabwareOffsetX) > LABWARE_OFFSET_DISPLAY_THRESHOLD
-      ? nestedLabwareOffsetX
-      : 0
-  const clampedLabwareOffsetY =
-    Math.abs(nestedLabwareOffsetY) > LABWARE_OFFSET_DISPLAY_THRESHOLD
-      ? nestedLabwareOffsetY
-      : 0
-  // transform to be applied to children which render within the labware interfacing surface of the module
-  const childrenTransform = `translate(${clampedLabwareOffsetX}, ${clampedLabwareOffsetY})`
-
```
Oh, heads up for this removal.

This code snapped a labware to the front-left corner of *the deck slot underneath the module* if the labware was within 2 mm of it.

I guess this makes sense for modules that provide a slot that's aligned with the underlying deck slot. The Temperature and Magnetic modules (our oldest modules) work like that. But it hasn't generally been true since then.

For simplicity, I am straight-up dropping this code. I think we should try to render the labware in the straightforward, physically-accurate way, because that's something that all of our code should be able to agree on how to do; and if we can stick with that, then we can avoid a lot of really problematic duplication.

Here's the effect of this removal: [before](https://github.com/user-attachments/assets/42530021-2b6b-441f-8a34-cbfcb13ebc82), [after](https://github.com/user-attachments/assets/a9c93fbd-6052-4e82-a409-8362997709e2).

SyntaxColoring replied:
Reconsidering this. Will post in Slack.

SyntaxColoring replied:
OK, spoke with design about this, we're good to drop this code.

If we ever want to add this behavior back in later, I don't think this is the best place to do it, because it won't affect places like `MoveLabwareOnDeck` that compute the absolute coordinates fully up-front, in TypeScript, instead of via a series of nested SVG transforms. I think we'd want to add an arg like `mode: "pretty" | "physicallyAccurate"` to `labwarePositioningMath.ts`.

SyntaxColoring committed:
Merge branch 'edge' into frontend_schema_3_labware_in_modules

Resolve conflicts in components/src/hardware-sim/BaseDeck/BaseDeck.tsx

SyntaxColoring committed:
Merge branch 'edge' into frontend_schema_3_labware_in_modules

Resolve conflicts in:

* app/src/organisms/Desktop/Devices/ProtocolRun/SetupLabware/SetupLabwareMap.tsx
* app/src/pages/Desktop/Protocols/ProtocolPreview/DeckViewDetails.tsx (Take `edge`, undoing the refactors in this PR. We'll redo them separately.)
* components/src/hardware-sim/BaseDeck/BaseDeck.tsx
* components/src/hardware-sim/Module/index.tsx

SyntaxColoring committed:
Update comment.

SyntaxColoring committed:
Fix a forgotten positioningMode="passThrough".

SyntaxColoring committed:
Fix missing alignment in some forgotten moduleChildren call sites.

These got missed because TypeScript couldn't trace the moduleChildren reference. They showed up in a vanilla text search.

SyntaxColoring committed:
Update story.

SyntaxColoring committed:
Merge branch 'edge' into frontend_schema_3_labware_in_modules

Resolve conflicts in:
* app/src/organisms/ErrorRecoveryFlows/shared/TwoColLwInfoAndDeck.tsx: This PR was updating some code to construct the `BaseDeck` `backgroundItems` prop. `edge` refactors the file to avoid using `backgroundItems` entirely. Take `edge`.
* app/src/organisms/InterventionModal/MoveLabwareInterventionContent.tsx: Same.

SyntaxColoring committed:
Post-merge lint fixup.

SyntaxColoring committed:
Update some mocks.
